Lib/urllib.py

```diff
@@ -203,7 +203,10 @@ def open(self, fullurl, data=None):
         name = 'open_' + urltype
         self.type = urltype
         name = name.replace('-', '_')
-        if not hasattr(self, name):
+
+        # bpo-35907: # disallow the file reading with the type not allowed
+        if not hasattr(self, name) or \
+            getattr(self, name) == self.open_local_file:
```

Comment:
Another fix is to rename the open_local_file() as _open_local_file(). But... In Python 2.7, there is a risk that the method is called directly in 3rd party code for whatever reason :-(

Reply:
Why do you compare function objects instead of simply ``name == "open_local_file"``?

Reply:
@tiran 
If someone rename `open_local_file()` method and forgets to modify `name == "open_local_file"`, it may be bypassed.
So I made the mitigation compare the object to raise syntax error. 
Should I do change back?

Reply:
I concur with Christian, please test directly the name, remove getattr().

Reply:
@vstinner 
I agree with you. I also considered renaming the `open_local_file()`. I think that this fix can guarantees availability to the developers who want to override this method. But I chose another fix to minimize side effects for third party codes.

Reply:
I have tested it. the mitigation has worked. or there is other reason for removing `getattr()`. If for just simplify. I will remove it.

Reply:
Please remove getattr().

Reply:
Yes. I will do it. I want suggest you other fix. 
```       
        if not hasattr(self, name):
            if proxy:
                return self.open_unknown_proxy(proxy, fullurl, data)
            else:
                return self.open_unknown(fullurl, data)
        if name == 'open_local_file': 
                raise IOError, ('url error', 'invalid url type', urltype)
        try:
```
I think that it is better to raise exception with explicit error message for developers to overriding `open_local_file()`.
they can easily recognize `local_file` is invalid type.
What do you think about it?
or should I use `open_unknown()`?
```
if not hasattr(self, name) or \
    name == 'open_local_file':
```

Reply:
```
# bpo-35907: disallow the file reading with the type not allowed 
if not hasattr(self, name) or name == 'open_local_file': ...
```

LGTM.

```
        if name == 'open_local_file': 
                raise IOError, ('url error', 'invalid url type', urltype)
```

That's wrong exception type, don't do that.

Reply:
I have changed it. Please review.

Reply:
I have made the requested changes; please review again.